lib/AST/ModuleNameLookup.cpp

```diff
@@ -18,93 +18,263 @@
+                     SmallVectorImpl<ValueDecl *> &localDecls) {
+    module->lookupValue(path, name, lookupKind, localDecls);
+  }
+};
+
+/// Encapsulates the work done for a recursive qualified lookup into a module
+/// to find all visible decls.
+class LookupVisibleDecls : public ModuleNameLookup<LookupVisibleDecls> {
+  using Super = ModuleNameLookup<LookupVisibleDecls>;
+  friend Super;
+
+  const NLKind lookupKind;
+
+public:
+  LookupVisibleDecls(LazyResolver *typeResolver, ModuleDecl *M,
+                     ResolutionKind resolutionKind, NLKind lookupKind)
+    : ModuleNameLookup(typeResolver, M, resolutionKind),
+      lookupKind(lookupKind) {}
+
+private:
+  using OverloadSetEntry =
+      std::pair<ResolutionKind, LookupByName::OverloadSetTy>;
+  using OverloadSetTy = llvm::DenseMap<DeclBaseName, OverloadSetEntry>;
+  static_assert(ResolutionKind() == ResolutionKind::Overloadable,
+                "Entries in NamedCanTypeSet should be overloadable initially");
 
-/// Updates \p overloads with the types of the given decls.
-///
-/// \returns true if all of the given decls are overloadable, false if not.
-static bool updateOverloadSet(CanTypeSet &overloads,
-                              ArrayRef<ValueDecl *> decls) {
-  for (auto result : decls) {
-    if (!isOverloadable(result))
+  /// Does \p VD conflict with the \p overloads we've already seen?
+  static bool isValidOverload(OverloadSetTy &overloads, const ValueDecl *VD) {
+    // Note: 'overloads' is not const because it's cheaper to create the empty
+    // set value under this name once and check it repeatedly.
+    const OverloadSetEntry &entry = overloads[VD->getBaseName()];
+    if (entry.first != ResolutionKind::Overloadable)
       return false;
-    if (!result->hasInterfaceType())
-      continue;
-    overloads.insert(result->getInterfaceType()->getCanonicalType());
+    return LookupByName::isValidOverload(entry.second, VD);
   }
-  return true;
-}
 
-/// Updates \p overloads with the types of the given decls.
-///
-/// \returns true, since there can always be more overloadable decls.
-static bool updateOverloadSet(NamedCanTypeSet &overloads,
-                              ArrayRef<ValueDecl *> decls) {
-  for (auto result : decls) {
-    auto &entry = overloads[result->getBaseName()];
-    if (!isOverloadable(result))
-      entry.first = ResolutionKind::Exact;
-    else if (!result->hasInterfaceType())
-      continue;
-    else
+  /// Updates \p overloads with the types of the given decls.
+  ///
+  /// \returns true, since there can always be more overloadable decls.
+  static bool updateOverloadSet(OverloadSetTy &overloads,
+                                ArrayRef<ValueDecl *> decls) {
+    for (auto result : decls) {
+      OverloadSetEntry &entry = overloads[result->getBaseName()];
+      if (!isOverloadable(result)) {
+        entry.first = ResolutionKind::Exact;
+        entry.second.clear();
```

Comment:
This `clear()` is the one new bit of behavior that got folded into this commit. It might mildly reduce memory usage during these lookups.